tests/suites/os/tests/secureboot/index.js

```diff
@@ -1,145 +1,285 @@
+			this.link
+		).then(() => this.worker.executeCommandInHostOS('reboot', this.link)
+		).then(() => this.test.resolves(
+			/* The below pattern is the expected output when the config file
+			 * fails the signature check and the fallback console is not enabled.
+			 *
+			 * This sequence of ANSI escape codes decodes as:
+			 *
+			 *	ESC[0m - reset all modes
+			 *	ESC[37m - white foreground
+			 *	ESC[40m - black background
+			 *
+			 *	https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797
+			 *
+			 * Otherwise, the output finishes the above background color escape
+			 * sequence, clears the screen, and shows 'loading Boot0003
+			 * "balenaOS"'
+			 */
+
+			this.waitForSerialOutput(/\x1B\[0m\x1B\[37m\x1B\[40m$/, 3), // eslint-disable-line no-control-regex
+			'Bootloader will not load configuration that fails signature verification',
+			{ todo: 'This needs reworked for GRUB 2.12' },
+		)
+		).then( () => this.resetDUT() );
+	};
+}
+
+class rpiSecureBoot extends secureBoot {
+	async isSecureBootSupported() {
+		/* If the partition is split and there are boot.img and boot.sig */
+		return true;
+	}
+
+	async readPrivateKey() {
+		await this.worker.executeCommandInHostOS(
+			['vcmailbox', '0x00030081', '40 40 0 8 0 0 0 0 0 0 0 0'],
+			this.link
+		).then((out) => {
+			const key = out
+				.replace(/0x/g, '') // remove 0x prefix
+				.trim()							// remove extra spaces
+				.split(/\s+/)				// split by whitespace
+				.slice(8, 16)				// take elements 8 to 15
+				.join('');          // concatenate
+			return key;
+		})
+	}
+
+	async readOTPReg(reg) {
+		await this.worker.executeCommandInHostOS(
+			['vcgencmd', 'otp_dump', '2>/dev/null'],
+			this.link
+		).then((out) => {
+				for ( const line of out.split('\n')) {
+					if (line.startsWith(reg)) {
+						return line.split(':')[1].trim();
 					}
-				} else {
-					test.comment('Secure boot is not supported by firmware');
 				}
+		})
+	}
+
+	async checkRSAkey() {
+		for (let reg = 47; reg <= 54; reg++) {
+			await readOTPReg(reg).then( (out) => {
+				if ( out && out != "00000000") {
+					return true
+				}
+			})
+		}
+		return false
+	}
+
+	async isSecureBootEnabled() {
+		// OTP is programmed with customer key digest and private key
+		if ( await checkRSAkey() ) {
+			const key = await readPrivateKey();
+			if ( key.replace(/0/g, '').trim().length > 0 ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	async testBootloaderIntegrity() {
+		/* Modify boot.sig to fail authentication */
+		await this.worker.executeCommandInHostOS(
+			`sed -d '1s/^.\(.*\)$/\1/' /mnt/rpi/boot.sig`,
+			this.link,
+		).then( () => this.worker.executeCommandInHostOS('reboot', this.link)
+		).then( () => this.test.resolves(
+			this.waitForSerialOutput(/TBC/),
+			'Bootloader will not load configuration that fails signature verification',
+		)
+		).then( () => this.resetDUT() );
+	}
+
+	async testBootloaderConfigIntegrity() {
```

Comment:
To be honest, I wonder if we should just refactor these functions to use async/await syntax rather than `.then()` /callbacks, as I personally find it a lot clearer to tell what is going on and in what order - i.e:

```js
await this.worker.executeCommandInHostOS(
			"tmpDir=$(mktemp -d) && mount -o loop /mnt/rpi/boot.img ${tmpDir} && sed -i 's/$/#comment/' ${tmpDir}/config.txt && umount ${tmpDir}",
			this.link)
await this.worker.executeCommandInHostOS('reboot', this.link)
await this.test.resolves(
			this.waitForSerialOutput(/TBC/),
			'Bootloader will not load configuration that fails signature verification',
		)
await this.resetDUT();
```